pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,154 @@
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// RemoveOSDs purges a list of OSDs from the cluster
+func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdsToRemove []string) error {
+
+	// Generate the ceph config for running ceph commands similar to the operator
+	if err := writeCephConfig(context, clusterInfo); err != nil {
+		return errors.Wrap(err, "failed to write the ceph config")
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		logger.Infof("removing OSD %q", osdIDStr)
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
+			logger.Errorf("failed to remove OSD %d. %v", osdID, err)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
+
+	// Check if the OSD is down
+	osdDump, err := client.GetOSDDump(context, clusterInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
+		id64, err := osdStatus.OSD.Int64()
+		if err != nil {
+			continue
+		}
+		id := int(id64)
+
+		logger.Debugf("validating status of osd.%d", id)
+
+		status, _, err := osdDump.StatusByID(int64(id))
+		if err != nil {
+			return err
+		}
+
+		const upStatus int64 = 1
+		var osdDown bool
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.", id)
+			osdDown = false
+			continue
+
+		}
+
+		logger.Debugf("osd.%d is marked 'DOWN'", id)
+
+		// Get the host where the OSD is found
+		hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+		if !osdDown {
+			return fmt.Errorf("OSD %d cannot be removed unless it is 'down'", osdID)
+		}
+
+		// Mark the OSD as out. If it is not down, it is expected this command will fail, in which case
+		// we expect to abort purging the OSD
+		output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
+		if err != nil {
+			return errors.Wrapf(err, fmt.Sprintf("cannot remove the osd. %s", output))
+		}
+
+		// Remove the OSD deployment
+		deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+		logger.Infof("removing the OSD deployment %q", deploymentName)
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if !kerrors.IsNotFound(err) {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+
+		// If running on a PVC, delete the OSD prepare job
+		prepareJobName := fmt.Sprintf("rook-ceph-osd-prepare-example-deviceset-0-#{osdID}")
```

Comment:
The osdID is not part of the osd prepare job name. We will need to lookup the osd prepare job differently, such as seeing that it has the same PVC as the OSD deployment that is being deleted.